homeassistant/util/network.py

```diff
@@ -86,6 +87,20 @@ def is_ipv6_address(address: str) -> bool:
     return True
 
 
+def is_host_valid(host: str) -> bool:
+    """Check if a given string is an IP address or valid hostname."""
+    if is_ip_address(host):
+        return True
+    if len(host) > 255:
+        return False
+    if re.match(r"^[0-9\.]+$", host):  # reject invalid IPv4
+        return False
+    if host.endswith("."):  # dot at the end is correct
+        host = host[:-1]
+    allowed = re.compile(r"(?!-)[A-Z\d\-]{1,63}(?<!-)$", re.IGNORECASE)
```

Comment:
If you make the `re`s constants you won't have to compile them every time the function runs

Reply:
But then it will compile even if we don't use it at all. I don't even know which is better :)

Reply:
Thats a good point. Its not complex enough that it probably matters, and we won't likely run it enough

Reply:
You could hide it behind an @cache but the cache overhead might be more than the compile anyways

Reply:
I think that most of the checks will end up not reaching the regular expressions, since IP addresses are mainly used.